File: trees_and_graphs/maximum_difference_between_node_and_ancestor.py
import treenode

class Solution:
    def maxAncestorDiff1(self, root: treenode.TreeNode) -> int:
        if not root:
            return 0
        
        # Record the required max difference
        self.result = 0

        def helper(node, cur_max, cur_min):
            if not node:
                return
            
            # Update result
            self.result = max(self.result,
                              abs(cur_max - node.val),
                              abs(cur_min - node.val))
            
            # Update the max and min
            cur_max = max(cur_max, node.val)
            cur_min = min(cur_min, node.val)

            helper(node.left, cur_max, cur_min)
            helper(node.right, cur_max, cur_min)

        helper(root, root.val, root.val)
        return self.result

File: trees_and_graphs/maximum_difference_between_node_and_ancestor.java
class Solution
{
    // Record the required maximum difference
    int result = 0;

    public void helper1(TreeNode node, int curMax, int curMin)
    {
        if (node == null)
        {
            return;
        }

        // Update result
        int possibleResult = Math.max(Math.abs(curMax - node.val),
                                      Math.abs(curMin - node.val));
        result = Math.max(result, possibleResult);

        // Update the max and min
        curMax = Math.max(curMax, node.val);
        curMin = Math.min(curMin, node.val);

        helper1(node.left, curMax, curMin);
        helper1(node.right, curMax, curMin);
    }

    public int maxAncestorDiff1(TreeNode root)
    {
        if (root == null)
        {
            return 0;
        }

        result = 0;
        helper1(root, root.val, root.val);
        return result;
    }
}